src/Authorization.AspNetCore/DefaultAuthorizationErrorMessageBuilder.cs

```diff
@@ -0,0 +1,114 @@
+            AppendFailureHeader(error, operationType);
+
+            if (result.Failure != null)
+            {
+                foreach (var requirement in result.Failure.FailedRequirements)
+                {
+                    AppendFailureLine(error, requirement);
+                }
+            }
+
+            return error.ToString();
+        }
+
+        private string GetOperationType(OperationType? operationType)
+        {
+            return operationType switch
+            {
+                OperationType.Query => "query",
+                OperationType.Mutation => "mutation",
+                OperationType.Subscription => "subscription",
+                _ => "operation",
+            };
+        }
+
+        /// <inheritdoc />
+        public virtual void AppendFailureHeader(StringBuilder error, OperationType? operationType)
+        {
+            error
+                .Append("You are not authorized to run this ")
+                .Append(GetOperationType(operationType))
+                .Append('.');
+        }
+
+        /// <inheritdoc />
+        public virtual void AppendFailureLine(StringBuilder error, IAuthorizationRequirement authorizationRequirement)
+        {
+            error.AppendLine();
+
+            switch (authorizationRequirement)
+            {
+                case ClaimsAuthorizationRequirement claimsAuthorizationRequirement:
+                    error.Append("Required claim '");
+                    error.Append(claimsAuthorizationRequirement.ClaimType);
+                    if (!claimsAuthorizationRequirement.AllowedValues.Any())
+                    {
+                        error.Append("' is not present.");
+                    }
+                    else
+                    {
+                        error.Append("' with any value of '");
+                        error.Append(string.Join(", ", claimsAuthorizationRequirement.AllowedValues));
+                        error.Append("' is not present.");
+                    }
+                    break;
+
+                case DenyAnonymousAuthorizationRequirement _:
+                    error.Append("The current user must be authenticated.");
+                    break;
+
+                case NameAuthorizationRequirement nameAuthorizationRequirement:
+                    error.Append("The current user name must match the name '");
+                    error.Append(nameAuthorizationRequirement.RequiredName);
+                    error.Append("'.");
+                    break;
+
+                case OperationAuthorizationRequirement operationAuthorizationRequirement:
+                    error.Append("Required operation '");
+                    error.Append(operationAuthorizationRequirement.Name);
+                    error.Append("' was not present.");
+                    break;
+
+                case RolesAuthorizationRequirement rolesAuthorizationRequirement:
+                    if (!rolesAuthorizationRequirement.AllowedRoles.Any())
```

Comment:
Again?

Reply:
I explained that check for null is needed to avoid NRE.

Reply:
> because in it's constructor an exception is thrown if attempting to create a RolesAuthorizationRequirement with AllowedRoles null

Did you need to decompile the source code to check this? In any case this is the current implementation that may vary, so it is not wise to remove the check.